test_vm/tests/datacap_tests.rs

```diff
@@ -0,0 +1,188 @@
+
+use fil_actor_verifreg::{AllocationRequest, AllocationRequests};
+use fil_actors_runtime::cbor::serialize;
+use fil_actors_runtime::runtime::policy_constants::MINIMUM_VERIFIED_ALLOCATION_SIZE;
+use fil_actors_runtime::runtime::Policy;
+use fil_actors_runtime::test_utils::make_piece_cid;
+use fil_actors_runtime::{DATACAP_TOKEN_ACTOR_ADDR, VERIFIED_REGISTRY_ACTOR_ADDR};
+use fvm_shared::error::ExitCode;
+use test_vm::util::{apply_code, apply_ok, create_accounts, create_miner};
+use test_vm::VM;
+
+use fil_actor_datacap::{Method as DataCapMethod, MintParams};
+use frc46_token::token::types::TransferFromParams;
+
+/* Mint a token for client and transfer it to a receiver, exercising error cases */
+#[test]
+fn datacap_transfer_scenario() {
+    let policy = Policy::default();
+    let store = MemoryBlockstore::new();
+    let mut v = VM::new_with_singletons(&store);
+    let addrs = create_accounts(&v, 3, TokenAmount::from_whole(10_000));
+    let (client, operator, owner) = (addrs[0], addrs[1], addrs[2]);
+
+    // need to allocate to an actual miner actor to pass verifreg receiver hook checks
+    let seal_proof = RegisteredSealProof::StackedDRG32GiBV1P1;
+    let (maddr, _) = create_miner(
+        &mut v,
+        owner,
+        owner,
+        seal_proof.registered_window_post_proof().unwrap(),
+        TokenAmount::from_whole(1_000),
+    );
+
+    let data_cap_amt = TokenAmount::from_whole(
+        MINIMUM_VERIFIED_ALLOCATION_SIZE + MINIMUM_VERIFIED_ALLOCATION_SIZE / 2,
+    );
+    let mint_params = MintParams { to: client, amount: data_cap_amt, operators: vec![operator] };
+
+    // cannot mint from non-verifreg
+    apply_code(
+        &v,
+        operator,
+        DATACAP_TOKEN_ACTOR_ADDR,
+        TokenAmount::zero(),
+        DataCapMethod::Mint as u64,
+        mint_params.clone(),
+        ExitCode::USR_FORBIDDEN,
+    );
+
+    // mint datacap for client
+    apply_ok(
+        &v,
+        VERIFIED_REGISTRY_ACTOR_ADDR,
+        DATACAP_TOKEN_ACTOR_ADDR,
+        TokenAmount::zero(),
+        DataCapMethod::Mint as u64,
+        mint_params,
+    );
+
+    let alloc = AllocationRequest {
+        provider: maddr,
+        data: make_piece_cid("datacap-test-alloc".as_bytes()),
+        size: PaddedPieceSize(MINIMUM_VERIFIED_ALLOCATION_SIZE as u64),
+        term_min: policy.minimum_verified_allocation_term,
+        term_max: policy.maximum_verified_allocation_term,
+        expiration: v.get_epoch() + policy.maximum_verified_allocation_expiration,
+    };
+    let transfer_from_params = TransferFromParams {
+        to: VERIFIED_REGISTRY_ACTOR_ADDR,
+        from: client,
+        amount: TokenAmount::from_whole(MINIMUM_VERIFIED_ALLOCATION_SIZE),
+        operator_data: serialize(
+            &AllocationRequests { allocations: vec![alloc.clone()], extensions: vec![] },
+            "operator data",
+        )
+        .unwrap(),
+    };
+    let clone_params = |x: &TransferFromParams| -> TransferFromParams {
```

Comment:
Do you wanna file an issue in the token lib to derive Clone for all these types?

Reply:
Thanks will do